### PR TITLE
[MRG+1] plot_date() after axhline() doesn't rescale axes

### PR DESCRIPTION
**Pull request by Yasaman-Mah**
The only reason that plotting axhline first changes behaviour of plotting datetime is that it sets ignore_existing_data_limits to false and then when plotting datetime data, default datetime data limits are taken into consideration.
Unfortunately this fix would affect y data limits too. I think a better solution would be to have two variables ignore_existing_x_limits and ignore_existing_y_limits, but I am not sure if adding more detail and making it slightly more complicated just because of axhline makes sense. 

ref #7742

### CODE REVIEW

File: lib/matplotlib/tests/test_axes.py [hunk 1]
```diff
@@ -4972,3 +4972,21 @@ def test_bar_single_height():
     ax.bar(range(4), 1)
     # Check that a horizontal chart with one width works
     ax.bar(0, 1, bottom=range(4), width=1, orientation='horizontal')
+
+def test_datetime_axhline_same_axes():
```
Comment (phobson) [inline lib/matplotlib/tests/test_axes.py]: The pep8 checker is probably going to want 2 blank lines between functions

Comment (phobson) [inline lib/matplotlib/tests/test_axes.py]: Looks like you'll also need to trim trailing whitespace: 
https://travis-ci.org/matplotlib/matplotlib/jobs/208246492#L1683

**Comment by NelleV** [comment]
Next time, please add a meaningful title to the pull requests. The reference to the ticket should be in in the description of the PR (else it is non clickable).

**Comment by Shang-Jia** [comment]
This solution will always reset the existing_data_limits so on the following execution
plot() -> sets data limits
axhline() -> allows data limits to be set again
plot() -> sets data limits again, even if data limits are smaller than what was previously plotted
see testcase and proposed solution in #8210

**Comment by Yasaman-Mah** [comment]
@Shang-Jia  This is a preliminary solution, I submitted this pull request to get some feedback from matplotlib team before going on with another solution.

**Comment by NelleV** [comment]
This looks good to me. Thanks for the patch and sorry it took so long to get feedback!

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -725,6 +725,7 @@ def axhline(self, y=0, xmin=0, xmax=1, **kwargs):
         trans = self.get_yaxis_transform(which='grid')
         l = mlines.Line2D([xmin, xmax], [y, y], transform=trans, **kwargs)
         self.add_line(l)
+        self.ignore_existing_data_limits = True
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: Can you add a symmetric change to `axvline` ?

**Comment by Yasaman-Mah** [comment]
This solution has a side effect, it will result in inconsistent behaviour when setting y axis limits. For example:
```
`import matplotlib.pyplot as plt

fig, axs = plt.subplots(2, 1)

axs[0].axvline(1)
axs[0].plot([2, 3], [2, 3])

axs[1].plot([2, 3], [2, 3])
axs[1].axvline(1)

plt.show()`
```
will result in following plots:
screenshot from 2017-03-15 20 36 29
which are not wrong but inconsistent. 
However, I believe currently there is no way to avoid this unless we add one more field to axes object or make more substantial changes to how limits are set.

File: lib/matplotlib/axes/_axes.py (hunk 2 above)
Comment (Yasaman-Mah) [inline lib/matplotlib/axes/_axes.py]: I have made a similar change to `axvline`. Please see my comment below approving the changes.

**Comment by ibnIrshad** [comment]
@Yasaman-Mah It's good you pointed that out. However I still think that this PR is a useful contribution and should be merged. 

This PR fixes #7742 (which produces clearly wrong output), and although it does introduce a side effect, the side effect is not "wrong output", but only slightly inconsistent for a certain specific order of input. And as you pointed out, it would require MUCH more work to fix this inconsistency, because introducing another symmetric flag "ignore_y_limits" is a significant architectural change.

Hence I would ask @NelleV @tacaswell to review this PR again. Can we go ahead with this fix?

**Comment by tacaswell** [comment]
What are dates (which are just input with units) behaving differently than plain numbers?

I am still not sure I understand what the `self.ignore_existing_data_limits` flag does and when it gets set/unset.

The code and the figure above seem out of sync with each other, but I would not call that a 'side effect'.  IT looks like the bug from #7742 (order dependent limits) has just been moved to different (and probably more common) case.

**Comment by jklymak** [comment]
Can this PR be resurrected or closed?

**Comment by jklymak** [comment]
I'm closing for now - but feel free to re-open if someone wants to champion this PR.  Thanks for the contribution.

**Pull request closed without merging**